sfe/sfe.go

```diff
@@ -0,0 +1,301 @@
+func init() {
+	tmplPages = template.Must(template.New("pages").ParseFS(dynamicFS, "templates/layout.html", "pages/*"))
+}
+
+// SelfServiceFrontEndImpl provides all the logic for Boulder's selfservice
+// frontend web-facing interface, i.e., a portal where a subscriber can unpause
+// their account. Its methods are primarily handlers for HTTPS requests for the
+// various non-ACME functions.
+type SelfServiceFrontEndImpl struct {
+	ra rapb.RegistrationAuthorityClient
+	sa sapb.StorageAuthorityReadOnlyClient
+
+	log blog.Logger
+	clk clock.Clock
+
+	// requestTimeout is the per-request overall timeout.
+	requestTimeout time.Duration
+
+	// unpausePublicKey should be derived from a seed value shared by the SFE
+	// and WFE. It is used to validate incoming JWT signatures on the unpause
+	// endpoint.
+	unpausePublicKey ed25519.PublicKey
+}
+
+// NewSelfServiceFrontEndImpl constructs a web service for Boulder
+func NewSelfServiceFrontEndImpl(
+	stats prometheus.Registerer,
+	clk clock.Clock,
+	logger blog.Logger,
+	requestTimeout time.Duration,
+	rac rapb.RegistrationAuthorityClient,
+	sac sapb.StorageAuthorityReadOnlyClient,
+	unpausePublicKey ed25519.PublicKey,
+) (SelfServiceFrontEndImpl, error) {
+	sfe := SelfServiceFrontEndImpl{
+		log:              logger,
+		clk:              clk,
+		requestTimeout:   requestTimeout,
+		ra:               rac,
+		sa:               sac,
+		unpausePublicKey: unpausePublicKey,
+	}
+
+	return sfe, nil
+}
+
+// Handler returns an http.Handler that uses various functions for various
+// non-ACME-specified paths. Each endpoint should have a corresponding HTML
+// page that shares the same name as the endpoint.
+func (sfe *SelfServiceFrontEndImpl) Handler(stats prometheus.Registerer, oTelHTTPOptions ...otelhttp.Option) http.Handler {
+	m := http.NewServeMux()
+
+	sfs, _ := fs.Sub(staticFS, "static")
+	staticAssetsHandler := http.StripPrefix("/static/", http.FileServerFS(sfs))
+
+	m.Handle("GET /static/", staticAssetsHandler)
+	m.HandleFunc("/", sfe.Index)
+	m.HandleFunc("GET /build", sfe.BuildID)
+	m.HandleFunc(unpausePath, sfe.Unpause)
+
+	return measured_http.New(m, sfe.clk, stats, oTelHTTPOptions...)
+}
+
+// renderTemplate takes an HTML template instantiated by the SFE init() and an
+// optional dynamicData which are rendered and served back to the client via the
+// response writer.
+func renderTemplate(w http.ResponseWriter, subTmpl string, dynamicData any) {
+	if len(subTmpl) == 0 {
+		http.Error(w, "Template page does not exist", http.StatusInternalServerError)
+		return
+	}
+
+	err := tmplPages.ExecuteTemplate(w, subTmpl, dynamicData)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// Index is the homepage of the SFE
+func (sfe *SelfServiceFrontEndImpl) Index(response http.ResponseWriter, request *http.Request) {
+	if request.Method != http.MethodGet && request.Method != http.MethodHead {
+		response.Header().Set("Access-Control-Allow-Methods", "GET, HEAD")
```

Comment:
[Access-Control-Allow-Methods](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods) is used as part of a [preflight request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods), which is part of CORS.

We don't intend to allow any CORS requests to any of these endpoints, so we can remove the Access-Control-Allow-Methods headers.